### DISCUSSION TITLE
Running os.proc fails with "Unknown ansi-escape" error

### DISCUSSION THREAD

**Discussion by jk-1** [lihaoyi/os-lib#155] [Q&A]
I try to run following ssh-ping command with os.proc but it fails with "Unknown ansi-escape" error. 

    val res = os.proc("ssh-ping", "-q", "-W", "2", "-c", "1", "-p", "22", "myhost").call(check=true, stdin=os.Pipe, stderr=os.Pipe) 
    java.lang.IllegalArgumentException: Unknown ansi-escape [0;93m at index 16 inside string cannot be parsed into an fansi.Str
      fansi.ErrorMode$Throw$.handle(Fansi.scala:419)
      fansi.ErrorMode$Throw$.handle(Fansi.scala:407)
      fansi.Str$.apply(Fansi.scala:272)
      pprint.Renderer.$anonfun$rec$32(Renderer.scala:147)
      scala.collection.Iterator$$anon$10.next(Iterator.scala:461)
      scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:514)
      pprint.Truncated.hasNext(Truncated.scala:80)
      pprint.Renderer.rec(Renderer.scala:45)
      pprint.Renderer.rec(Renderer.scala:42)
      pprint.Renderer.rec(Renderer.scala:42)
      pprint.PPrinter.tokenize(PPrinter.scala:104)
      ammonite.repl.FullReplAPI$Internal.print(FullReplAPI.scala:106)
      ammonite.repl.FullReplAPI$Internal.print$(FullReplAPI.scala:61)
      ammonite.repl.FullReplAPI$$anon$1.print(FullReplAPI.scala:34)

I also tried to add various environment variables like
`env=Map("COLOR" -> "off", "TERM" -> "xterm-256color")` or `"TERM" -> "xterm-mono"` or `"TERM" -> "dumb"` etc without any success.

On bash terminal output is this:

    ssh-ping -q -W 2 -c 1 -p 22 myhost
    SSHPING myhost

    --- myhost ping statistics ---
    1 requests transmitted, 0 requests received, 100% request loss

ssh-ping command has no switch to turn off colors.

    ssh-ping --help

    Usage: ssh-ping [OPTIONS] [user@]hostname

    OPTIONS:

        -4             Use IPv4 only
        -6             Use IPv6 only
        -c count       Stop after sending <count> request packets
        -F configfile  Specifies an alternative per-user configuration file.
                       If a configuration file is given on the command line,
                       the system-wide configuration file ( /etc/ssh/ssh_config ) will be ignored.
                       The default for the per-user configuration file is ~/.ssh/config.
        -h             Show this message
        -i interval    Wait <interval> seconds between sending each request.
                       The default is 1 second.
        -l user        Try login with <user> as username. The default is $USER.
        -D             Print timestamp (unix time + microseconds as in gettimeofday) before each line
        -H             Print timestamp (human readable) before each line
        -W timeout     Time to wait for a response, in seconds
        -p port        Port to connect to on the remote host.
                       This can be specified on a per-host basis in the configuration file.
        -q             Quiet output.
                       Nothing is displayed except the summary lines at startup time and when finished
        -v             Verbose output

ssh-ping is part of ssh-tools package.

Question: Is this error in os.proc or maybe in fansi or can I somehow configure os.proc or even fansi to accept unknown ansi-escapes? How can I circumvent this error and e.g. force mono chrome output?

If an Ubuntu MATE terminal can handle these ansi-escapes, then maybe also os.proc / fansi should be able to handle them without crashing.

Thank you for your support!

I use Ubuntu 20.04, Ammonite Repl 2.4.0 (Scala 2.12.13 Java 11.0.18)

**Comment by SethTisue** [comment]
Looks like the error is coming from fansi; maybe there's something about it at https://github.com/com-lihaoyi/fansi/issues ?

**Comment by MarylynnGutermuth0d** [comment]
It seems that the error is caused by the fansi library used by Ammonite's pretty-printing (pprint) feature, which is trying to parse an unknown ANSI escape sequence in the output of the ssh-ping command.

One possible solution is to disable pretty-printing in the Ammonite REPL by setting the pprintEnable property to false before executing the os.proc command. For example:
import pprint.Config

pprint.Config.pprintEnable() = false

val res = os.proc("ssh-ping", "-q", "-W", "2", "-c", "1", "-p", "22", "myhost").call(check=true, stdin=os.Pipe, stderr=os.Pipe)

This should disable pretty-printing and prevent the fansi library from interfering with the output of the ssh-ping command.

Alternatively, you can try setting the TERM environment variable to a non-color terminal, such as xterm-mono or dumb, before executing the command. For example:

val res = os.proc("ssh-ping", "-q", "-W", "2", "-c", "1", "-p", "22", "myhost").call(check=true, stdin=os.Pipe, stderr=os.Pipe, env=Map("TERM" -> "xterm-mono"))

This should tell the ssh-ping command to output in a non-color format that can be processed correctly by the fansi library.

**Reply by SethTisue** [reply]
this is ChatGPT generated spam

**Reply by jk-1** [reply]
> One possible solution is to disable pretty-printing in the Ammonite REPL by setting the pprintEnable property to false before executing the os.proc command. For example: import pprint.Config

Which versions are you using because I get this error:

```
jk@tii:~/Downloads/PPrint-master$ amm
Loading...
Welcome to the Ammonite Repl 2.4.0 (Scala 2.12.13 Java 11.0.18)
jk-PPrint-master@ import $ivy.`com.lihaoyi::pprint:0.7.0` 
import $ivy.$                          

jk-PPrint-master@  

jk-PPrint-master@ import pprint._ 
import pprint._

jk-PPrint-master@  

jk-PPrint-master@ import pprint.Config 
cmd1.sc:149: object Config is not a member of package pprint
possible cause: maybe a semicolon is missing before `object Config'?
import pprint.Config
       ^
Compilation Failed

```

> 
> Alternatively, you can try setting the TERM environment variable to a non-color terminal, such as xterm-mono or dumb, before executing the command. For example:
> 
> val res = os.proc("ssh-ping", "-q", "-W", "2", "-c", "1", "-p", "22", "myhost").call(check=true, stdin=os.Pipe, stderr=os.Pipe, env=Map("TERM" -> "xterm-mono"))

Unfortunately this does not fix the problem, output here:

```
jk-PPrint-master@ val res = os.proc("ssh-ping", "-q", "-W", "2", "-c", "1", "-p", "22", "myhost").call(check=true, stdin=os.Pipe, stderr=os.Pipe, env=Map("TERM" -> "xterm-mono")) 
Exception in thread "main" java.lang.IllegalArgumentException: Unknown ansi-escape [0;93m at index 56 inside string cannot be parsed into an fansi.Str
	at fansi.ErrorMode$Throw$.handle(Fansi.scala:419)
	at fansi.ErrorMode$Throw$.handle(Fansi.scala:407)
	at fansi.Str$.apply(Fansi.scala:272)
	at fansi.Str$.implicitApply(Fansi.scala:227)
	at ammonite.repl.Repl$.$anonfun$showException$1(Repl.scala:308)
	at scala.collection.immutable.List.map(List.scala:293)
	at ammonite.repl.Repl$.showException(Repl.scala:304)
	at ammonite.repl.Repl$.handleRes(Repl.scala:262)
	at ammonite.repl.Repl.loop$1(Repl.scala:219)
	at ammonite.repl.Repl.run(Repl.scala:227)
	at ammonite.Main.$anonfun$run$1(Main.scala:232)
	at scala.Option.getOrElse(Option.scala:189)
	at ammonite.Main.run(Main.scala:220)
	at ammonite.MainRunner.$anonfun$runRepl$1(Main.scala:425)
	at ammonite.MainRunner.watchLoop(Main.scala:406)
	at ammonite.MainRunner.runRepl(Main.scala:425)
	at ammonite.Main$.main0(Main.scala:331)
	at ammonite.Main$.main(Main.scala:279)
	at ammonite.Main.main(Main.scala)

```

**Reply by SethTisue** [reply]
> Which versions are you using

It isn't using any versions. It's an AI spewing random nonsense.

**Comment by JessieLeddonR** [comment]
The error you're seeing is coming from the pprint library, which is used by Ammonite REPL to pretty-print values in the console. The library is not able to parse the ANSI escape code that is being printed by ssh-ping.

To work around this issue, you can try disabling colors in the output of ssh-ping. Looking at the ssh-ping help output, it seems that the -q (quiet) option might be enough to achieve this. You can try running the command with -q and see if it produces output that pprint is able to handle.

Alternatively, you can try redirecting the output of ssh-ping to a file, and then reading the contents of the file into your Scala program. This way, you can process the output without having to worry about the ANSI escape codes. Here's an example of how you can do this:
import java.io.File

val outputFile = File.createTempFile("ssh-ping-", ".txt")
val command = Seq("ssh-ping", "-q", "-W", "2", "-c", "1", "-p", "22", "myhost") #> outputFile
val exitCode = command.!
val output = os.read(outputFile)
println(output)

In this example, we create a temporary file to store the output of ssh-ping. We then use the #> operator from the os-lib library to redirect the output of ssh-ping to the file. Finally, we read the contents of the file into a Scala string using the os.read function.

Note that the ! operator is used to execute the command and wait for it to finish. If you don't want to wait for the command to finish, you can use the run method instead.

**Reply by SethTisue** [reply]
this is ChatGPT generated spam

**Reply by SethTisue** [reply]
I'm getting rate limited in my effort to report these accounts as spammers. Could someone else watching this repo try it...?

**Reply by jk-1** [reply]
> To work around this issue, you can try disabling colors in the output of ssh-ping. Looking at the ssh-ping help output, it seems that the -q (quiet) option might be enough to achieve this. You can try running the command with -q and see if it produces output that pprint is able to handle.

As you can see from my original example code, I was already using the -q, but even when using the -q option the ssh-ping output contains ansi-escapes that pprint is not able to handle. 
 
> Alternatively, you can try redirecting the output of ssh-ping to a file, and then reading the contents of the file into your Scala program. This way, you can process the output without having to worry about the ANSI escape codes. 

Writing into temp files and reading them might work as a temporary Q&D solution to workaround this issue. However it is not a preferred long term solution, this ssh-ping is just a simple to reproduce example where os.proc / pprint fails. I remember getting same type of exceptions when using os.proc also earlier with some other commands but was too busy to report those at that time.
